rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,275 @@
+            if (locationOfName(qn) == just(defined)) return true;
+        case variableBecomes(n, _):
+            if (locationOfName(n) == just(defined)) return true;
+    }
+
+    return false;
+}
+
+default bool isImplicitDefinition(start[Module] _, rel[loc, loc] _, Define _) = false;
+
+void checkCausesCapture(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = {def | Define def <- newDefs, isImplicitDefinition(m, tm.useDef, def)};
+    set[Define] newNameExplicitDefs = {def | Define def <- newDefs, !isImplicitDefinition(m, tm.useDef, def)};
+
+    // Will this rename turn an implicit declaration of `newName` into a use of a current declaration?
+    rel[loc use, loc oldDef, loc newDef] implicitDeclBecomesUseOfCurrentDecl =
+        {<nD.defined, nD.defined, cD> | Define nD <- newNameImplicitDefs
+                                      , <loc cS, loc cD> <- currentDefs<scope, defined>
+                                      , isContainedIn(nD.defined, cS)
+        };
+
+    // Will this rename turn a use of <oldName> into a use of an existing declaration of <newName> (shadowing)?
+    rel[loc use, loc oldDef, loc newDef] explicitDeclShadowsCurrentDecl =
+        {<cU, cD.defined, nD.defined> | Define nD <- newNameExplicitDefs
+                                      , loc cU <- currentUses
+                                      , Define cD <- getDefines(tm, cU)
+                                      , isContainedIn(cU, nD.scope)
+                                      , isStrictlyContainedIn(nD.scope, cD.scope)
+        };
+
+    throwIfNotEmpty(CausesCapture, implicitDeclBecomesUseOfCurrentDecl + explicitDeclShadowsCurrentDecl);
+}
+
+void checkEditableDefinitions(set[Define] currentDefs) {
+    bool isEditableLocation(loc l) = l.scheme in {"file", "cwd", "home", "memory", "project"};
+    throwIfNotEmpty(DeclarationsNotEditable, {def | loc def <- currentDefs<defined>, !isEditableLocation(def)});
+}
+
+void checkLegalRename(TModel tm, start[Module] m, loc cursorLoc, set[Define] currentDefs, set[loc] currentUses, str newName) {
+    set[Define] newNameDefs = {def | def <- tm.defines, def.id == newName};
+    try {
+        checkEditableDefinitions(currentDefs);
+        checkLegalName(newName);
+        checkCausesDoubleDeclaration(currentDefs, newNameDefs);
+        checkCausesCapture(tm, m, currentDefs, currentUses, newNameDefs);
+    } catch IllegalRenameReason r: {
+        throw IllegalRename(cursorLoc, r);
+    }
+    checkUnsupported(m.src, currentDefs);
+}
+
+void checkUnsupported(loc moduleLoc, set[Define] defsToRename) {
+    Maybe[str] isUnsupportedDefinition(Define _: <scope, id, _, functionId(), _, _>) = just("Global function definitions might span multiple modules; unsupported for now.")
+        when moduleLoc == scope; // function is defined in module scope
+    default Maybe[str] isUnsupportedDefinition(Define _) = nothing();
```

Comment:
wouldn't it be easier to mark what we do support? instead of all the cases we don't? Like I'm not seeing module names or grammar names for example.

Reply:
In geneal, `findDeclarationName` will take care of what we do and don't support. In this case however, we need some more analysis to see if we really support this definition.

Reply:
Done!